phaser/pipeline.py

```diff
@@ -129,8 +153,51 @@ def run(self):
         next_source = self.source
         for phase in self.phase_instances:
             destination = self.get_destination(phase)
-            phase.run(source=next_source, destination=destination)
-            next_source = destination  # for next phase in chain
+            self.run_phase(phase, next_source, destination)
+            next_source = destination
+
+    def run_phase(self, phase, source, destination):
+        logger.info(f"Loading input from {source} for {phase.name}")
+        data = self.load(source)
+        phase.load_data(data)
+        results = phase.run()
+        self.save(results, destination)
+        self.save_extra_outputs()
+        logger.info(f"{phase.name} saved output to {destination}")
+        if self.context.has_errors():
+            raise PipelineErrorException(f"Phase '{phase.name}' failed with {len(self.context.errors.keys())} errors.")
+
+    def save_extra_outputs(self):
+        for item in self.context.outputs:
+            # Since context is passed from Phase to Phase, only save the new ones with to_save=True
+            if item.to_save:
+                filename = self.working_dir / f"{item.name}.csv"
+                if os.path.exists(filename):
+                    raise PhaserException(f"Output with name '{filename}' exists.  Aborting before overwrite.")
+                pd.DataFrame(item.data).to_csv(filename, index=False, na_rep="NULL")
```

Comment:
It is probably worth pulling this function out into a `utils.write_csv()` function so we can isolate the dependency on Pandas, put all I/O in one place, and have one place where we control how we write csv files.